File: Utils/src/main/java/com/sheryv/util/Observable.java
package com.sheryv.util;

import javax.annotation.Nullable;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class Observable<T> {
  
  private final List<ChangeListener<T>> listeners = Collections.synchronizedList(new ArrayList<>());
  
  public Observable(T initial) {
    this.value = initial;
  }
  
  public Observable() {
  
  }
  
  public void addListener(ChangeListener<T> listener) {
    this.listeners.add(listener);
  }
  
  public synchronized void removeListener(ChangeListener<T> listener) {
    try {
      this.listeners.remove(listener);
    } catch (Exception e) {
      return;
    }
  }
  
  private T value;
  
  @Nullable
  public T getValue() {
    return this.value;
  }
  
  public void set(@Nullable T newValue) {
    T oldValue = this.value;
    this.value = newValue;
    firePropertyChange(oldValue, newValue);
  }
  
  public void firePropertyChange(T oldValue, T newValue) {
    List<ChangeListener<T>> copy = null;
    boolean run = false;
    
    synchronized (this) {
      if (oldValue == null || newValue == null || !oldValue.equals(newValue)) {
        copy = new ArrayList<>(listeners);
        run = true;
      }
    }
    
    if (run) {
      for (ChangeListener<T> changeListener : copy) {
        changeListener.onChange(oldValue, newValue);
      }
    }
  }
}

File: Utils/src/main/kotlin/com/sheryv/util/Observable.kt
package com.sheryv.util

import java.util.*

class Observable<T>(initial: T? = null) {
  private val listeners = Collections.synchronizedList(ArrayList<ChangeListener<T?>>())
  
  
  fun addListener(listener: ChangeListener<T?>) {
    listeners.add(listener)
  }
  
  @Synchronized
  fun removeListener(listener: ChangeListener<T?>) {
    try {
      listeners.remove(listener)
    } catch (e: Exception) {
      return
    }
  }
  
  var value: T? = initial
    set(value) {
      val oldValue: T? = field
      field = value
      firePropertyChange(oldValue, value)
    }
  
  
  fun firePropertyChange(oldValue: T?, newValue: T?) {
    var copy: List<ChangeListener<T?>>? = null
    var run = false
    synchronized(this) {
      if (oldValue == null || newValue == null || oldValue != newValue) {
        copy = ArrayList(listeners)
        run = true
      }
    }
    if (run) {
      for (changeListener in copy!!) {
        changeListener.onChange(oldValue, newValue)
      }
    }
  }
}
